### ISSUE TITLE
FreqAI  backtesting and hyperopt Error

### ISSUE THREAD

**Issue by BuddhiLW** [freqtrade/freqtrade#8981]
## Describe your environment

  * Operating system: Ubuntu
  * Python Version:  (`Python 3.11.2`)
  * CCXT version:  (`ccxt==3.1.44`)
  * Freqtrade Version:  (`freqtrade 2023.6`)
  
Note: All issues other than enhancement requests will be closed without further comment if the above template is deleted or not filled out.

## Describe the problem:
Running:

`freqtrade hyperopt --config config_train-not-cheap-coins-USDT.json --hyperopt-loss SortinoHyperOptLoss --strategy BLW_train4 -e 10000 --spaces buy sell roi trailing --timerange=20230101-20230401 --freqaimodel CatboostClassifier`

Gives:

```
2023-07-30 20:57:34,185 - freqtrade.freqai.freqai_interface - WARNING - Training XMR/USDT raised exception ValueError. Message: With n_samples=0, test_size=0.25 and train_size=None, the resulting train set will be empty. Adjust any of the aforementioned parameters., skipping.
```

Train strategy:
```python
# pragma pylint: disable=missing-docstring, invalid-name, pointless-string-statement
# isort: skip_file
# --- Do not remove these libs ---
import logging
from typing import Dict

import numpy as np  # noqa
import pandas as pd  # noqa
from pandas import DataFrame
from freqtrade.strategy.interface import IStrategy
from freqtrade.strategy import (
    BooleanParameter,
    CategoricalParameter,
    DecimalParameter,
    IStrategy,
    IntParameter,
    merge_informative_pair,
)
from functools import reduce

# --------------------------------
# Add your lib to import here
import talib.abstract as ta
import freqtrade.vendor.qtpylib.indicators as qtpylib

# from technical import qtpylib

logger = logging.getLogger(__name__)

# This class is a sample. Feel free to customize it.
class BLW_train4(IStrategy):
    buy_adx = DecimalParameter(20, 60, decimals=1, default=32.7, space="buy")
    buy_rsi = IntParameter(15, 40, default=22, space="buy")
    # buy_adx_enabled_less = BooleanParameter(default=False, space="buy")
    buy_adx_enabled_more = BooleanParameter(default=True, space="buy")
    # buy_rsi_enabled_less = CategoricalParameter([True, False], default=True, space="buy")
    buy_rsi_enabled_more = CategoricalParameter([True, False], default=True, space="buy")
    # buy_stoch_enabled = BooleanParameter(default=False, space="buy")
    # buy_stoch = IntParameter(0, 35, default=20, space="buy")
    buy_doji = IntParameter(30, 80, default=50, space="buy")
    buy_doji_enabled = BooleanParameter(default=False, space="buy")
    # buy_CDLHANGINGMAN = IntParameter(0, 100, default=50, space="buy")
    # buy_CDLHANGINGMAN_enabled = BooleanParameter(default=False, space="buy")
    # buy_CDLSHOOTINGSTAR = IntParameter(0, 100, default=50, space="buy")
    # buy_CDLSHOOTINGSTAR_enabled = BooleanParameter(default=True, space="buy")
    # buy_CDLGRAVESTONEDOJI = IntParameter(0, 100, default=50, space="buy")
    # buy_CDLGRAVESTONEDOJI_enabled = BooleanParameter(default=True, space="buy")
    # buy_CDLDARKCLOUDCOVER = IntParameter(0, 100, default=50, space="buy")
    # buy_CDLDARKCLOUDCOVER_enabled = BooleanParameter(default=False, space="buy")
    # buy_CDLEVENINGDOJISTAR = IntParameter(0, 100, default=50, space="buy")
    # buy_CDLEVENINGDOJISTAR_enabled = BooleanParameter(default=False, space="buy")
    # buy_CDLEVENINGSTART = IntParameter(0, 100, default=50, space="buy")
    # buy_CDLEVENINGSTART_enabled = BooleanParameter(default=True, space="buy")
    buy_trigger = CategoricalParameter(
        [
            "bb_lower1",
            "bb_lower2",
            "bb_lower3",
            "bb_lower4",
            # "macd_cross_signal",
            # "dragon_fly"  # buy_doji
            # "bb_percent",
            # "bb_width",
            # "stoch",
        ],
        default="bb_lower1",
        space="buy",
    )

    sell_adx = DecimalParameter(60, 90, decimals=1, default=77.4, space="sell")
    sell_rsi = IntParameter(50, 100, default=66, space="sell")
    sell_adx_enabled_less = BooleanParameter(default=True, space="sell")
    sell_rsi_enabled_less = CategoricalParameter([True, False], default=False, space="sell")
    sell_adx_enabled_more = BooleanParameter(default=False, space="sell")
    sell_rsi_enabled_more = CategoricalParameter([True, False], default=True, space="sell")
    sell_stoch_enabled = BooleanParameter(default=False, space="sell")
    sell_stoch = IntParameter(60, 100, default=70, space="sell")
    sell_spinningtop = IntParameter(-100, 0, default=-50, space="sell")
    sell_spinningtop_enabled = BooleanParameter(default=True, space="sell")
    sell_trigger = CategoricalParameter(
        [
            "spinning_top",
            "bb_lower1",
            "bb_lower2",
            # "bb_lower3",
            # "bb_lower4",
            "bb_mid1",
            # "bb_mid2",
            # "bb_mid3",
            # "bb_mid4",
            "bb_upper1",
            "bb_upper2",
            # "bb_upper3",
            # "bb_upper4",
            "macd_cross_signal",
            # "bb_percent",
            # "bb_width",
            "stoch",
        ],
        default="bb_mid1",
        space="sell",
    )
    # sell_trigger = CategoricalParameter(
    #     ["bb_mid1", "bb_mid4", "macd_cross_signal"], default="bb_mid1", space="sell"
    # )
    # sell_trigger = CategoricalParameter(
    #     ["bb_upper1", "bb_upper4", "macd_cross_signal"], default="bb_upper1", space="sell"
    # )

    # Strategy interface version - allow new iterations of the strategy interface.
    # Check the documentation or the Sample strategy to get the latest version.
    # INTERFACE_VERSION = 2

    # Minimal ROI designed for the strategy.
    # This attribute will be overridden if the config file contains "minimal_roi".
    # ROI table:
    minimal_roi = {"0": 0.399, "188": 0.11, "411": 0.069, "519": 0}

    # Stoploss:
    stoploss = -0.05

    # Trailing stop:
    trailing_stop = True
    trailing_stop_positive = 0.086
    trailing_stop_positive_offset = 0.092
    trailing_only_offset_is_reached = True

    # Optimal ticker interval for the strategy.
    timeframe = "5m"

    # Run "populate_indicators()" only for new candle.
    process_only_new_candles = True

    # These values can be overridden in the "ask_strategy" section in the config.
    use_exit_signal = True
    exit_profit_only = False
    ignore_roi_if_sell_signal = False
    # Number of candles the strategy requires before producing valid signals
    startup_candle_count: int = 30

    # Optional order type mapping.
    order_types = {
        "entry": "limit",
        "exit": "limit",
        "emergency_exit": "market",
        "force_entry": "market",
        "force_exit": "market",
        "stoploss": "market",
        "stoploss_on_exchange": False,
        "stoploss_on_exchange_interval": 60,
        "stoploss_on_exchange_limit_ratio": 0.99,
    }

    process_only_new_candles = False

    plot_config = {
        "main_plot": {
            "dragon_fly": {"color": "pink"},
            # "tema": {},
            # "sar": {"color": "white"},
        },
        "subplots": {
            "VOLUME": {
                "volume": {"color": "orange"},
            },
            "ADX": {
                "adx": {"color": "black"},
            },
            "MACD": {
                "macd": {"color": "blue"},
                "macdsignal": {"color": "orange"},
            },
            "RSI": {
                "rsi": {"color": "red"},
            },
            # "Up_or_down": {
            #     "&s-up_or_down": {"color": "green"},
            # },
        },
    }

    std_dev_multiplier_buy = CategoricalParameter(
        [0.75, 1, 1.25, 1.5, 1.75], default=1.25, space="buy", optimize=True
    )
    std_dev_multiplier_sell = CategoricalParameter(
        [0.75, 1, 1.25, 1.5, 1.75], space="sell", default=1.25, optimize=True
    )

    def informative_pairs(self):
        """
        Define additional, informative pair/interval combinations to be cached from the exchange.
        These pair/interval combinations are non-tradeable, unless they are part
        of the whitelist as well.
        For more information, please consult the documentation
        :return: List of tuples in the format (pair, interval)
            Sample: return [("ETH/USDT", "5m"),
                            ("BTC/USDT", "15m"),
                            ]
        """
        return []

    def feature_engineering_expand_all(
        self, dataframe: DataFrame, period: int, metadata: Dict, **kwargs
    ) -> DataFrame:
        """
        *Only functional with FreqAI enabled strategies*
        This function will automatically expand the defined features on the config defined
        `indicator_periods_candles`, `include_timeframes`, `include_shifted_candles`, and
        `include_corr_pairs`. In other words, a single feature defined in this function
        will automatically expand to a total of
        `indicator_periods_candles` * `include_timeframes` * `include_shifted_candles` *
        `include_corr_pairs` numbers of features added to the model.

        All features must be prepended with `%` to be recognized by FreqAI internals.

        Access metadata such as the current pair/timeframe with:

        `metadata["pair"]` `metadata["tf"]`

        More details on how these config defined parameters accelerate feature engineering
        in the documentation at:

        https://www.freqtrade.io/en/latest/freqai-parameter-table/#feature-parameters

        https://www.freqtrade.io/en/latest/freqai-feature-engineering/#defining-the-features

        :param dataframe: strategy dataframe which will receive the features
        :param period: period of the indicator - usage example:
        :param metadata: metadata of current pair
        dataframe["%-ema-period"] = ta.EMA(dataframe, timeperiod=period)
        """

        dataframe["%-rsi-period"] = ta.RSI(dataframe, timeperiod=period)
        dataframe["%-mfi-period"] = ta.MFI(dataframe, timeperiod=period)
        dataframe["%-adx-period"] = ta.ADX(dataframe, timeperiod=period)
        dataframe["%-sma-period"] = ta.SMA(dataframe, timeperiod=period)
        dataframe["%-ema-period"] = ta.EMA(dataframe, timeperiod=period)

        bollinger = qtpylib.bollinger_bands(
            qtpylib.typical_price(dataframe), window=period, stds=2.2
        )
        dataframe["bb_lowerband-period"] = bollinger["lower"]
        dataframe["bb_middleband-period"] = bollinger["mid"]
        dataframe["bb_upperband-period"] = bollinger["upper"]

        dataframe["%-bb_width-period"] = (
            dataframe["bb_upperband-period"] - dataframe["bb_lowerband-period"]
        ) / dataframe["bb_middleband-period"]
        dataframe["%-close-bb_lower-period"] = dataframe["close"] / dataframe["bb_lowerband-period"]

        dataframe["%-roc-period"] = ta.ROC(dataframe, timeperiod=period)

        dataframe["%-relative_volume-period"] = (
            dataframe["volume"] / dataframe["volume"].rolling(period).mean()
        )

        return dataframe

    def feature_engineering_expand_basic(
        self, dataframe: DataFrame, metadata: Dict, **kwargs
    ) -> DataFrame:
        """
        *Only functional with FreqAI enabled strategies*
        This function will automatically expand the defined features on the config defined
        `include_timeframes`, `include_shifted_candles`, and `include_corr_pairs`.
        In other words, a single feature defined in this function
        will automatically expand to a total of
        `include_timeframes` * `include_shifted_candles` * `include_corr_pairs`
        numbers of features added to the model.

        Features defined here will *not* be automatically duplicated on user defined
        `indicator_periods_candles`

        All features must be prepended with `%` to be recognized by FreqAI internals.

        Access metadata such as the current pair/timeframe with:

        `metadata["pair"]` `metadata["tf"]`

        More details on how these config defined parameters accelerate feature engineering
        in the documentation at:

        https://www.freqtrade.io/en/latest/freqai-parameter-table/#feature-parameters

        https://www.freqtrade.io/en/latest/freqai-feature-engineering/#defining-the-features

        :param dataframe: strategy dataframe which will receive the features
        :param metadata: metadata of current pair
        dataframe["%-pct-change"] = dataframe["close"].pct_change()
        dataframe["%-ema-200"] = ta.EMA(dataframe, timeperiod=200)
        """
        dataframe["%-pct-change"] = dataframe["close"].pct_change()
        dataframe["%-raw_volume"] = dataframe["volume"]
        dataframe["%-raw_price"] = dataframe["close"]
        return dataframe

    def feature_engineering_standard(
        self, dataframe: DataFrame, metadata: Dict, **kwargs
    ) -> DataFrame:
        """
        *Only functional with FreqAI enabled strategies*
        This optional function will be called once with the dataframe of the base timeframe.
        This is the final function to be called, which means that the dataframe entering this
        function will contain all the features and columns created by all other
        freqai_feature_engineering_* functions.

        This function is a good place to do custom exotic feature extractions (e.g. tsfresh).
        This function is a good place for any feature that should not be auto-expanded upon
        (e.g. day of the week).

        All features must be prepended with `%` to be recognized by FreqAI internals.

        Access metadata such as the current pair with:

        `metadata["pair"]`

        More details about feature engineering available:

        https://www.freqtrade.io/en/latest/freqai-feature-engineering

        :param dataframe: strategy dataframe which will receive the features
        :param metadata: metadata of current pair
        usage example: dataframe["%-day_of_week"] = (dataframe["date"].dt.dayofweek + 1) / 7
        """
        dataframe["%-day_of_week"] = dataframe["date"].dt.dayofweek
        dataframe["%-hour_of_day"] = dataframe["date"].dt.hour
        return dataframe

    def set_freqai_targets(self, dataframe: DataFrame, metadata: Dict, **kwargs) -> DataFrame:
        """
        *Only functional with FreqAI enabled strategies*
        Required function to set the targets for the model.
        All targets must be prepended with `&` to be recognized by the FreqAI internals.

        Access metadata such as the current pair with:

        `metadata["pair"]`

        More details about feature engineering available:

        https://www.freqtrade.io/en/latest/freqai-feature-engineering

        :param dataframe: strategy dataframe which will receive the targets
        :param metadata: metadata of current pair
        usage example: dataframe["&-target"] = dataframe["close"].shift(-1) / dataframe["close"]
        """
        dataframe["&-s_close"] = (
            dataframe["close"]
            .shift(-self.freqai_info["feature_parameters"]["label_period_candles"])
            .rolling(self.freqai_info["feature_parameters"]["label_period_candles"])
            .mean()
            / dataframe["close"]
            - 1
        )

        # Classifiers are typically set up with strings as targets:
        # df['&s-up_or_down'] = np.where( df["close"].shift(-100) >
        #                                 df["close"], 'up', 'down')

        # If user wishes to use multiple targets, they can add more by
        # appending more columns with '&'. User should keep in mind that multi targets
        # requires a multioutput prediction model such as
        # freqai/prediction_models/CatboostRegressorMultiTarget.py,
        # freqtrade trade --freqaimodel CatboostRegressorMultiTarget

        # df["&-s_range"] = (
        #     df["close"]
        #     .shift(-self.freqai_info["feature_parameters"]["label_period_candles"])
        #     .rolling(self.freqai_info["feature_parameters"]["label_period_candles"])
        #     .max()
        #     -
        #     df["close"]
        #     .shift(-self.freqai_info["feature_parameters"]["label_period_candles"])
        #     .rolling(self.freqai_info["feature_parameters"]["label_period_candles"])
        #     .min()
        # )

        return dataframe

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:

        # FreqAI
        dataframe = self.freqai.start(dataframe, metadata, self)

        for val in self.std_dev_multiplier_buy.range:
            dataframe[f"target_roi_{val}"] = (
                dataframe["&-s_close_mean"] + dataframe["&-s_close_std"] * val
            )

        for val in self.std_dev_multiplier_sell.range:
            dataframe[f"sell_roi_{val}"] = (
                dataframe["&-s_close_mean"] - dataframe["&-s_close_std"] * val
            )
        # return dataframe
        # ADX
        dataframe["adx"] = ta.ADX(dataframe)

        # MACD
        macd = ta.MACD(dataframe)
        dataframe["macd"] = macd["macd"]
        dataframe["macdsignal"] = macd["macdsignal"]
        dataframe["macdhist"] = macd["macdhist"]

        # RSI
        dataframe["rsi"] = ta.RSI(dataframe)

        # Bollinger Bands
        bollinger1 = qtpylib.bollinger_bands(qtpylib.typical_price(dataframe), window=20, stds=1)
        dataframe["bb_lower1"] = bollinger1["lower"]
        dataframe["bb_mid1"] = bollinger1["mid"]
        dataframe["bb_upper1"] = bollinger1["upper"]

        bollinger2 = qtpylib.bollinger_bands(qtpylib.typical_price(dataframe), window=20, stds=2)
        dataframe["bb_lower2"] = bollinger2["lower"]
        dataframe["bb_mid2"] = bollinger2["mid"]
        dataframe["bb_upper2"] = bollinger2["upper"]

        bollinger3 = qtpylib.bollinger_bands(qtpylib.typical_price(dataframe), window=20, stds=3)
        dataframe["bb_lower3"] = bollinger3["lower"]
        dataframe["bb_mid3"] = bollinger3["mid"]
        dataframe["bb_upper3"] = bollinger3["upper"]

        bollinger4 = qtpylib.bollinger_bands(qtpylib.typical_price(dataframe), window=30, stds=4)
        dataframe["bb_lower4"] = bollinger4["lower"]
        dataframe["bb_mid4"] = bollinger4["mid"]
        dataframe["bb_upper4"] = bollinger4["upper"]

        # dataframe["bb_percent"] = (dataframe["close"] - dataframe["bb_lower1"]) / (
        #     dataframe["bb_upper1"] - dataframe["bb_lower1"]
        # )
        # dataframe["bb_width"] = (dataframe["bb_upper1"] - dataframe["bb_lower1"]) / dataframe[
        #     "bb_mid1"
        # ]

        # stochastic
        stoch_fast = ta.STOCHF(dataframe, 5, 3, 0, 3, 0)
        dataframe["fastd"] = stoch_fast["fastd"]
        dataframe["fastk"] = stoch_fast["fastk"]

        # CDL Dragron Fly
        dataframe["dragon_fly"] = ta.CDLDRAGONFLYDOJI(dataframe)

        dataframe["CDLHANGINGMAN"] = ta.CDLHANGINGMAN(dataframe)
        # # Shooting Star: values [0, 100]
        dataframe["CDLSHOOTINGSTAR"] = ta.CDLSHOOTINGSTAR(dataframe)
        # # Gravestone Doji: values [0, 100]
        dataframe["CDLGRAVESTONEDOJI"] = ta.CDLGRAVESTONEDOJI(dataframe)
        # # Dark Cloud Cover: values [0, 100]
        dataframe["CDLDARKCLOUDCOVER"] = ta.CDLDARKCLOUDCOVER(dataframe)
        # # Evening Doji Star: values [0, 100]
        dataframe["CDLEVENINGDOJISTAR"] = ta.CDLEVENINGDOJISTAR(dataframe)
        # # Evening Star: values [0, 100]
        dataframe["CDLEVENINGSTAR"] = ta.CDLEVENINGSTAR(dataframe)

        # CDL Spinning Top
        dataframe["spinning_top"] = ta.CDLSPINNINGTOP(dataframe)  # values [0, -100, 100]
        return dataframe

    def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        conditions = []

        # GUARDS AND TRENDS
        if self.buy_adx_enabled_more.value:
            conditions.append(dataframe["adx"] > self.buy_adx.value)
        if self.buy_rsi_enabled_more.value:
            conditions.append(dataframe["rsi"] > self.buy_rsi.value)
        if self.buy_doji_enabled.value:
            conditions.append(dataframe["dragon_fly"] > self.buy_doji.value)

        # TRIGGERS
        if self.buy_trigger.value == "bb_lower2":
            conditions.append(dataframe["close"] < dataframe["bb_lower2"])

        # Check that volume is not 0
        conditions.append(dataframe["volume"] > 0)
        conditions.append(dataframe["do_predict"] == 1)
        conditions.append(
            dataframe["&-s_close"] > dataframe[f"target_roi_{self.std_dev_multiplier_buy.value}"]
        )
        # conditions.append(dataframe["&-s_close"] == "up")
        # conditions.append(dataframe["&s-up_or_down"] == "up")

        if conditions:
            dataframe.loc[reduce(lambda x, y: x & y, conditions), "buy"] = 1

        return dataframe

    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        """
        Based on TA indicators, populates the sell signal for the given dataframe
        :param dataframe: DataFrame populated with indicators
        :param metadata: Additional information, like the currently traded pair
        :return: DataFrame with sell column
        """
        conditions = []

        # GUARDS AND TRENDS
        if self.sell_adx_enabled_less.value:
            conditions.append(dataframe["adx"] < self.sell_adx.value)
        if self.sell_adx_enabled_more.value:
            conditions.append(dataframe["adx"] > self.sell_adx.value)
        if self.sell_rsi_enabled_more.value:
            conditions.append(dataframe["rsi"] > self.sell_rsi.value)
        # if self.sell_rsi_enabled_less.value:
        #     conditions.append(dataframe["rsi"] < self.sell_rsi.value)
        if self.sell_spinningtop_enabled.value:
            conditions.append(dataframe["spinning_top"] < self.sell_spinningtop.value)
        # if self.sell_stoch_enabled.value:
        # if self.sell_stoch_enabled.value:
        #     conditions.append(
        #         (dataframe["fastk"] > self.sell_stoch.value)
        #         and (dataframe["fastd"] > self.sell_stoch.value)
        #     )

        # TRIGGERS

        if self.sell_trigger.value == "bb_lower1":
            conditions.append(dataframe["bb_lower1"] < self.sell_spinningtop.value)
        if self.sell_trigger.value == "bb_lower2":
            conditions.append(dataframe["bb_lower2"] < self.sell_spinningtop.value)
        if self.sell_trigger.value == "bb_lower3":
            conditions.append(dataframe["bb_lower3"] < self.sell_spinningtop.value)
        if self.sell_trigger.value == "bb_mid1":
            conditions.append(dataframe["bb_mid1"] < self.sell_spinningtop.value)
        if self.sell_trigger.value == "bb_upper1":
            conditions.append(dataframe["bb_upper1"] < self.sell_spinningtop.value)
        if self.sell_trigger.value == "bb_upper2":
            conditions.append(dataframe["bb_upper2"] < self.sell_spinningtop.value)
        if self.sell_trigger.value == "macd_cross_signal":
            conditions.append(qtpylib.crossed_below(dataframe["macd"], dataframe["macdsignal"]))
        if self.sell_trigger.value == "spinning_top":
            conditions.append(qtpylib.crossed_above(dataframe["fastk"], dataframe["fastd"]))

        # Check that volume is not 0
        conditions.append(dataframe["volume"] > 0)
        conditions.append(
            dataframe["&-s_close"]
            < dataframe[f"sell_roi_{self.std_dev_multiplier_sell.value}"] * 0.25
        )

        if conditions:
            dataframe.loc[reduce(lambda x, y: x & y, conditions), "exit_long"] = 1

        return dataframe
```

### Observed Results:

  * What happened? `error with backtesting and hyperopt`
  * What did you expect to happen? `working backtesting and hyperopt`

### Relevant code exceptions or logs

Note: Please copy/paste text of the messages, no screenshots of logs please.

  ```
2023-07-30 20:57:34,185 - freqtrade.freqai.freqai_interface - WARNING - Training XMR/USDT raised exception ValueError. Message: With n_samples=0, test_size=0.25 and train_size=None, the resulting train set will be empty. Adjust any of the aforementioned parameters., skipping.
Traceback (most recent call last):
  File "/home/lw/PP/freqtrade/freqtrade/freqai/freqai_interface.py", line 354, in start_backtesting
    self.model = self.train(dataframe_train, pair, dk)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/freqai/base_models/BaseClassifierModel.py", line 53, in train
    dd = dk.make_train_test_datasets(features_filtered, labels_filtered)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/freqai/data_kitchen.py", line 156, in make_train_test_datasets
    ) = train_test_split(
        ^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/.env/lib/python3.11/site-packages/sklearn/model_selection/_split.py", line 2448, in train_test_split
    n_train, n_test = _validate_shuffle_split(
                      ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/.env/lib/python3.11/site-packages/sklearn/model_selection/_split.py", line 2126, in _validate_shuffle_split
    raise ValueError(
ValueError: With n_samples=0, test_size=0.25 and train_size=None, the resulting train set will be empty. Adjust any of the aforementioned parameters.
2023-07-30 20:57:34,186 - freqtrade.freqai.freqai_interface - INFO - Saving metadata to disk.
2023-07-30 20:57:34,186 - freqtrade - ERROR - Fatal exception!
Traceback (most recent call last):
  File "/home/lw/PP/freqtrade/freqtrade/main.py", line 42, in main
    return_code = args['func'](args)
                  ^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/commands/optimize_commands.py", line 107, in start_hyperopt
    hyperopt.start()
  File "/home/lw/PP/freqtrade/freqtrade/optimize/hyperopt.py", line 553, in start
    self.prepare_hyperopt_data()
  File "/home/lw/PP/freqtrade/freqtrade/optimize/hyperopt.py", line 466, in prepare_hyperopt_data
    preprocessed = self.advise_and_trim(data)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/optimize/hyperopt.py", line 446, in advise_and_trim
    preprocessed = self.backtesting.strategy.advise_all_indicators(data)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/strategy/interface.py", line 1321, in advise_all_indicators
    return {pair: self.advise_indicators(pair_data.copy(), {'pair': pair}).copy()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/strategy/interface.py", line 1321, in <dictcomp>
    return {pair: self.advise_indicators(pair_data.copy(), {'pair': pair}).copy()
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/strategy/interface.py", line 1339, in advise_indicators
    return self.populate_indicators(dataframe, metadata)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/user_data/strategies/strategy_04_train.py", line 383, in populate_indicators
    dataframe = self.freqai.start(dataframe, metadata, self)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/freqai/freqai_interface.py", line 164, in start
    dk = self.start_backtesting(dataframe, metadata, self.dk, strategy)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/freqai/freqai_interface.py", line 371, in start_backtesting
    self.dd.save_metadata(dk)
  File "/home/lw/PP/freqtrade/freqtrade/freqai/data_drawer.py", line 430, in save_metadata
    dk.data["training_features_list"] = list(dk.data_dictionary["train_features"].columns)
                                             ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
KeyError: 'train_features'

  ```

**Comment by robcaulk** [comment]
Hello,

Firstly, please provide a *full* trace. Your log is missing key details and warning signs from freqai from earlier in your run.

Until you provide that, we are shooting in the dark. Here is my attempt:

Are you able to run backtesting alone (without hyperopt?)

If not, are you able to run backtesting using the example files from the source code (FreqaiExampleStrategy.py, config_freqai.example.json)?

If yes, can you run hyperopt on the example files?

cheers,

rob

**Comment by BuddhiLW** [comment]
I can't use FreqAI with hyperopt and backtesting even with sample strategies. But, the error that appears is one related to "Division by Zero". As I looked through other issues, it seems that the zero-division error relates to data availability - therefore, not related to this issue.

As soon as I get home, I will post the full trace. 

Thanks in advance,
Pedro

**Comment by robcaulk** [comment]
If FreqAI is warning you about having 0 data points, then you'll want to review our documentation:

https://www.freqtrade.io/en/stable/freqai-running/#downloading-data-to-cover-the-full-backtest-period

**Comment by xmatthias** [comment]
Closing this as i think your question has been answered.

Feel free to comment below if it's still unclear.

**Comment by BuddhiLW** [comment]
Hello! Sorry I didn't reply then. I couldn't reproduce the problem. But, I just did.

If you include these markets in your strategy
```
            "ETC/USDT:USDT",
            "INJ/USDT:USDT",
            "TOMO/USDT:USDT",
            "MATIC/USDT:USDT",
            "SOL/USDT:USDT",
            "BNB/USDT:USDT",
            "FLUX/USDT:USDT",
            "XMR/USDT:USDT",
            "XRP/USDT:USDT",
            "ETH/USDT:USDT",
```

And, backtest, or hyperopt etc, using a `timerange` including the interval `20230101-20230201`. Then, it will crash with the following error:

```
Traceback (most recent call last):
  File "/home/lw/PP/freqtrade/freqtrade/freqai/freqai_interface.py", line 354, in start_backtesting
    self.model = self.train(dataframe_train, pair, dk)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/freqai/base_models/BaseClassifierModel.py", line 41, in train
    features_filtered, labels_filtered = dk.filter_features(
                                         ^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/freqai/data_kitchen.py", line 247, in filter_features
    if (1 - len(filtered_df) / len(unfiltered_df)) > 0.1 and self.live:
            ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
ZeroDivisionError: division by zero
2023-09-27 16:37:39,331 - freqtrade.freqai.freqai_interface - INFO - Saving metadata to disk.
2023-09-27 16:37:39,331 - freqtrade - ERROR - Fatal exception!
Traceback (most recent call last):
  File "/home/lw/PP/freqtrade/freqtrade/main.py", line 42, in main
    return_code = args['func'](args)
                  ^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/commands/optimize_commands.py", line 107, in start_hyperopt
    hyperopt.start()
  File "/home/lw/PP/freqtrade/freqtrade/optimize/hyperopt.py", line 553, in start
    self.prepare_hyperopt_data()
  File "/home/lw/PP/freqtrade/freqtrade/optimize/hyperopt.py", line 466, in prepare_hyperopt_data
    preprocessed = self.advise_and_trim(data)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/optimize/hyperopt.py", line 446, in advise_and_trim
    preprocessed = self.backtesting.strategy.advise_all_indicators(data)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/strategy/interface.py", line 1321, in advise_all_indicators
    return {pair: self.advise_indicators(pair_data.copy(), {'pair': pair}).copy()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/strategy/interface.py", line 1321, in <dictcomp>
    return {pair: self.advise_indicators(pair_data.copy(), {'pair': pair}).copy()
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/strategy/interface.py", line 1339, in advise_indicators
    return self.populate_indicators(dataframe, metadata)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/templates/FreqaiBLW1-USDT-transition.py", line 222, in populate_indicators
    dataframe = self.freqai.start(dataframe, metadata, self)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/freqai/freqai_interface.py", line 164, in start
    dk = self.start_backtesting(dataframe, metadata, self.dk, strategy)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lw/PP/freqtrade/freqtrade/freqai/freqai_interface.py", line 371, in start_backtesting
    self.dd.save_metadata(dk)
  File "/home/lw/PP/freqtrade/freqtrade/freqai/data_drawer.py", line 430, in save_metadata
    dk.data["training_features_list"] = list(dk.data_dictionary["train_features"].columns)
                                             ~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
KeyError: 'train_features'
```

**Comment by xmatthias** [comment]
don't use hyperopt if backtesting ain't working - hyperopt is just "backtesting, but 100ds of times", really (that will simplify the exceptions, ...).

aside from that - can you post the full logs from startup to this error ? 
It sounds a bit like one of your pairs is missing data at the start - which currently ain't really supported (freqAI assumes all data is available for all pairs).

**Comment by LAMBOBLACK666** [comment]
a thousand times I checked a thousand times I made a clean installation a thousand times I unloaded all timeframes a thousand times I unloaded the data, I took only the default configs exclusively for the guide, it seems to me that I already know by heart what was written in the documentation BUT I didn't start anything.
KeyError: 'train_features'
neither training nor bektesting nor hooper wholesale. error in training occurs in the data and loss of all data. KeyError error occurs in bectesting and hooper wholesale: 'train _ features'
